lib/model/model.go

```diff
@@ -2497,16 +2513,25 @@ func (m *model) CommitConfiguration(from, to config.Configuration) bool {
 	// Pausing a device, unpausing is handled by the connection service.
 	fromDevices := from.DeviceMap()
 	toDevices := to.DeviceMap()
+	// Unique set of device IDs for which we'd like to keep pending folder entries
+	keepPendingFoldersFor := make(map[protocol.DeviceID]bool, len(toDevices))
```

Comment:
I experimented with the empty `struct{}` approach at first, but somehow ended up with a boolean.  I guess because I wanted to allow passing false values as well for different semantics, which are not needed anymore in this version.  Will revisit and probably change to empty structs.

The fact that we're passing a keep-list here is because that is all `CommitConfiguration()` can come up with unless we move the DB iterating code from `CleanPendingFolders()` over there.  Encapsulating that in `observed.go` was meant to not clutter the already complex model code with keyer functions and DB iterators.

Reply:
In CommitConfiguration you have both old and new devices, so you should be able to come up with a drop-list. However in the `onServe` code you don't have that information, so this might indeed have to stay how it is - which is fine, I'd just would have preferred it the other way around if possible (which it isn't).

Reply:
Revisiting this decision, I'm not so keen on using empty structs.  They cost zero memory, but something like

    keep[deviceID] = true

is much more readable IMO than

    keep[deviceID] = struct{}{}

Since the size of these maps is initally the same as the known devices list itself, I think the memory savings are totally negligible.  Also, writing helper functions for encapsulating this is really overkill, so I'd like to stick with the more readable version.

Reply:
I wonder if it might even be easiest to turn this on its head an work with *only* the boolean map, like this:
1. Insert all entries we positively want to keep with `true` values while iterating through the config sections anyway.
2. Overwrite entries we positively want to drop with `false` values.
3. Leave out anything we haven't encountered in the config, effectively the same as setting `false`.
4. Pass over all PendingFolder / PendingDevice entries *once*, doing map lookups along the way for each entry.

I don't know how efficiently LevelDB handles key iteration (entries are always stored sorted by key, right?).  But lookups in the DB (for `Delete(key)`) would supposedly be more expensive than lookups in a temporary RAM map, plus it saves the possible failure while generating each key.  What do you think?

This approach should work for both `onServe()` and `CommitConfiguration()`, though I haven't tried writing it down yet so not quite sure.

Reply:
Sorry for the noise, that would **not work** for all cases, because specifying PendingFolders for keeping / dropping requires both device and folder ID, so it would be a rather complex data structure again.

Inter-package APIs suck I guess ;-)

Reply:
The more I try to avoid these DB lookups / deletions, the code gets overly complex and almost impossible to understand.  Looking at the original approach where both pending, ignored and added {folders|devices} lived in the config, it seems so clear and concise.  I'm trying to come up with an approach based on that previous behavior.

For pending devices, it's rather easy: List ignored and added device IDs in a map of booleans, iterate over all pending devices and delete if they are in the map.

Pending folders in this branch's DB are keyed by folder first because I thought that would be more useful in the API.  In the config, they were grouped by device, so deleting the whole device section automatically drops the pending folder entries, which will definitely be more effort here.  Ensuring ignored and added folders are no longer pending takes very little code in the config, by basically doing what I've started with: For every shared or ignored folder / device combination, the pending folder entry is `delete()`d from a temporary map, whether it exists or not (the slightly expensive lookup part on the DB side).

None of that previously tried to take the state before / after a change into consideration, because it happened on each final config object state before `CommitConfiguration()` is even called.

This needs more thought to get done elegantly.  Changing the DB PendingFolderKey order to sort by device first might be part of the solution - at the cost of more expensive data structure conversions in the API call.  Then maybe try to mimic the PendingDevice handling from the config, and rethink what would be an efficient way to iterate only once over all PendingFolders with a pre-built drop decision map, which will be similar in complexity to what the `PendingFolders()` method now returns.  Thanks for your patience.